### PR TITLE
fix(slurm_ops): implement feedback from slurm-charms PR #35

### LINKED ISSUE #46: [Bug/_AptManager]: install fails for both slurmdbd and slurmrestd due to missing `/var/lib/slurm` directory
The `install(...)` method for `_AptManager` fails on the slurmdbd and slurmrestd charms because the deb package scripts for both those packages does not create the parent directory `/var/lib/slurm`:

https://github.com/charmed-hpc/hpc-libs/blob/fda80b391b08d5654f0e0d69f8522b7eca84b674/lib/charms/hpc_libs/v0/slurm_ops.py#L486

`parents=True` isn't provided to `mkdir(...)` so an error will be raised when installing either `slurmrestd` or `slurmdbd`.

The easiest fix is to just add an additional private method `_create_state_save_location` that creates that directory for each of the Slurm nodes that needs that directory to exist. It seems mostly relevant for `slurmdbd`, `slurmd`, and `slurmctld` as all these packages creates the Slurm user that needs to own these directories while `slurmrestd` does not.

### LINKED ISSUE #47: [Bug/_MungeManager]: incorrect service name for MUNGE when using `_AptManager`
As part of integration testing for https://github.com/charmed-hpc/slurm-charms/pull/35, turns out we have the incorrect service name for MUNGE if using the `_AptManager` implementation. The service name is `munge`, not `munged`:

```
unit-slurmctld-0: 09:27:19 ERROR unit.slurmctld/0.juju-log command ['systemctl', 'reload-or-restart', 'munged'] failed with message Failed to reload-or-restart munged.service: Unit munged.service not found.

unit-slurmctld-0: 09:27:19 ERROR unit.slurmctld/0.juju-log command ['systemctl', 'reload-or-restart', 'munged'] failed. stderr:
Failed to reload-or-restart munged.service: Unit munged.service not found.
```

This issue wasn't caught because the integration tests don't try to start or stop the MUNGE service, but the fix is relatively easy as the `_ServiceType` enum will need to have the service name changed to `munge`:

https://github.com/charmed-hpc/hpc-libs/blob/fda80b391b08d5654f0e0d69f8522b7eca84b674/lib/charms/hpc_libs/v0/slurm_ops.py#L177-L180

The additional work that will be needed here is that we'll need to change the service name on the Slurm snap to `munge` instead of `munged`. We'll just need to make sure that the service name `munge` doesn't clash with the `/snap/bin/munge` binary on the snap package.

### LINKED ISSUE #48: [Bug/SlurmrestdManager]: `_apply_overrides` raises an error when creating the `slurmrestd` user and group
`_apply_overrides` fails to create the `slurmrestd` user and group for the slurmrestd service because `subprocess.check_output(...)` raises an error when attempting to pass the gid `64031` as an integer rather than a string.

```pytb
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/./src/charm.py", line 105, in <module>
    main.main(SlurmrestdCharm)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/main.py", line 551, in main
    manager.run()
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/main.py", line 530, in run
    self._emit()
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/main.py", line 516, in _emit
    self.framework.reemit()
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/framework.py", line 870, in reemit
    self._reemit()
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/framework.py", line 950, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/./src/charm.py", line 55, in _on_install
    self._slurmrestd.install()
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/lib/charms/hpc_libs/v0/slurm_ops.py", line 483, in install
    self._apply_overrides()
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/lib/charms/hpc_libs/v0/slurm_ops.py", line 739, in _apply_overrides
    subprocess.check_output(["groupadd", "--gid", 64031, "slurmrestd"])
  File "/usr/lib/python3.10/subprocess.py", line 421, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.10/subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.10/subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.10/subprocess.py", line 1796, in _execute_child
    self.pid = _posixsubprocess.fork_exec(
TypeError: expected str, bytes or os.PathLike object, not int
```

Pretty easy fix since it looks like we just need to cast the gid and uid for slurmrestd to string. I guess mocking `subprocess.run(...)` in the slurm_ops unit tests allowed this `TypeError` to slip through? _pyright I thought you were my friend_ :sob:

### LINKED ISSUE #49: [Bug/_AptManager]: `StateSaveLocation` file permissions are too restrictive
The `slurmctld` service gets all busted and disgusted when starting up after install because the set permissions on the `StateSaveLocation` are too restrictive. The current permissions are `0o600` which means that only the user `slurm` can read/write to the directory:

https://github.com/charmed-hpc/hpc-libs/blob/fda80b391b08d5654f0e0d69f8522b7eca84b674/lib/charms/hpc_libs/v0/slurm_ops.py#L486

However, when we try to start the `slurmctld` service via `systemctl`, it fails with the error message below:

```log
× slurmctld.service - Slurm controller daemon
     Loaded: loaded (/lib/systemd/system/slurmctld.service; enabled; vendor preset: enabled)
    Drop-In: /etc/systemd/system/slurmctld.service.d
             └─10-slurmctld-nofile.conf
     Active: failed (Result: exit-code) since Wed 2024-11-13 14:44:41 UTC; 2s ago
       Docs: man:slurmctld(8)
    Process: 266530 ExecStart=/usr/sbin/slurmctld -D -s $SLURMCTLD_OPTIONS (code=exited, status=1/FAILURE)
   Main PID: 266530 (code=exited, status=1/FAILURE)
        CPU: 6ms
Nov 13 14:44:41 juju-058453-2 systemd[1]: Started Slurm controller daemon.
Nov 13 14:44:41 juju-058453-2 slurmctld[266530]: slurmctld: fatal: Incorrect permissions on state save loc: /var/lib/slurm/checkp>
Nov 13 14:44:41 juju-058453-2 systemd[1]: slurmctld.service: Main process exited, code=exited, status=1/FAILURE
Nov 13 14:44:41 juju-058453-2 systemd[1]: slurmctld.service: Failed with result 'exit-code'.
```

It looks that by slightly "loosening" the permissions to be `0o700` where the `slurm` user can read, write, and execute from `StateSaveLocation` is the preferred file permission mode. Looking at the `slurmd` package, it creates a `checkpoint` directory with the permissions set to `0o755`, so it looks like this is what the preferred file permission mode is for `StateSaveLocation`.

### LINKED ISSUE #50: [Bug/_AptManager]: munge must be explicitly installed as dependency for `slurmrestd` service
The `slurmrestd` Debian package does not declare MUNGE as a dependency, so while the other Slurm services implicitly bring in the `munge` package, `slurmrestd` does not. This causes the `slurmrestd` charm to explode when it handles the `SlurmctldAvailable` event because `mungectl` cannot find the _etc/munge_ directory as the directory has not been created by the `munge` package.

Relevant log output:

```pytb
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/./src/charm.py", line 105, in <module>
    main.main(SlurmrestdCharm)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/main.py", line 551, in main
    manager.run()
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/main.py", line 530, in run
    self._emit()
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/main.py", line 519, in _emit
    _emit_charm_event(self.charm, self.dispatcher.event_name)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/main.py", line 147, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/framework.py", line 348, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/framework.py", line 860, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/framework.py", line 950, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/src/interface_slurmctld.py", line 86, in _on_relation_changed
    self.on.slurmctld_available.emit(munge_key, slurm_conf)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/framework.py", line 348, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/framework.py", line 860, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/venv/ops/framework.py", line 950, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/./src/charm.py", line 75, in _on_slurmctld_available
    self._slurmrestd.munge.key.set(event.munge_key)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/lib/charms/hpc_libs/v0/slurm_ops.py", line 859, in set
    _mungectl("key", "set", stdin=key)
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/lib/charms/hpc_libs/v0/slurm_ops.py", line 174, in _mungectl
    return _call("mungectl", *args, stdin=stdin).stdout
  File "/var/lib/juju/agents/unit-slurmrestd-0/charm/lib/charms/hpc_libs/v0/slurm_ops.py", line 140, in _call
    raise SlurmOpsError(f"command {cmd} failed. stderr:\n{result.stderr}")
charms.hpc_libs.v0.slurm_ops.SlurmOpsError: command ['mungectl', 'key', 'set'] failed. stderr:
failed to write new munge key to /etc/munge/munge.key
```

The easiest fix is to explicitly install `munge` in the `_install_service` method along with the other packages that we install on all Slurm nodes:

https://github.com/charmed-hpc/hpc-libs/blob/fda80b391b08d5654f0e0d69f8522b7eca84b674/lib/charms/hpc_libs/v0/slurm_ops.py#L643

### LINKED ISSUE #51: [Bug/SlumrestdManager]: `/etc/slurm` directory must be created manually after installing `slurmrestd`
The `slurmrestd` package doesn't create the `/etc/slurm` directory even though `slurmrestd` requires a copy of the _slurm.conf_ configuration file to function correctly + contact the Slurm cluster controller.

This was something that was already handled by the _slurmrestd_ops_ module, so we just need to port the functionality over to `slurm_ops`:

https://github.com/charmed-hpc/slurm-charms/blob/9aee46d82fd8ddd4c150be0a809cd2438f2a6c72/charms/slurmrestd/src/slurmrestd_ops.py#L134-L135

### LINKED ISSUE #52: [Bug/SlurmrestdManager]: `libslurmfull.so` missing after installing `slurmrestd` service
`slurm-wlm-basic-plugins` must be explicitly installed. The `slurmrestd` service requires that `libslurmfull.so` is available on the host, but the deb package does not require the `slurm-wlm-basic-plugins` package which provides this shared library. This highlights that the `slurmrestd` deb package might need some additional work, but we'll see as we upgrade our base to Noble.

This was also something that was handled by the _slurmrestd_ops_ module, so we just need to port the functionality into `slurm_ops`:

https://github.com/charmed-hpc/slurm-charms/blob/9aee46d82fd8ddd4c150be0a809cd2438f2a6c72/charms/slurmrestd/src/slurmrestd_ops.py#L115

We should maybe also add a TODO comment that suggests we fix this missing dependency in the upstream `slurmrestd` package.

### PR DESCRIPTION
**Pull request by NucciTheBoss**
This PR implements the necessary fixes needed for `slurm_ops` to unblock https://github.com/charmed-hpc/slurm-charms/pull/35. The PR also bumps the `LIBPATCH` version to `8`.

### Related issues

* Fixes #52 
* Fixes #51 
* Fixes #50 
* Fixes #49 
* Fixes #48
* Fixes #47 
* Fixes #46

### CODE REVIEW

File: lib/charms/hpc_libs/v0/slurm_ops.py [hunk 1]
```diff
@@ -658,6 +655,22 @@ def _install_service(self) -> None:
...
+            to checkpoint runtime information should a service crash, and it
+            serves as the location where the JWT token used to generate user
+            access tokens is stored as well.
+        """
+        _logger.debug("creating slurm `StateSaveLocation` directory")
+        target = self.var_lib_path / "checkpoint"
+        target.mkdir(mode=0o755, parents=True, exist_ok=True)
+        self.var_lib_path.chmod(0o755)
+        shutil.chown(self.var_lib_path, "slurm", "slurm")
+        shutil.chown(target, "slurm", "slurm")
```
Comment (jedel1043) [inline lib/charms/hpc_libs/v0/slurm_ops.py:672]: Nit: Possibly changing the permissions for `var_lib_path` should be outside `_create_state_save_location`? Seems unrelated to creating the directory itself.

**Review by jedel1043 — APPROVED** [review]
Looks great! Just have a small nit that doesn't block merging :)

File: lib/charms/hpc_libs/v0/slurm_ops.py (hunk 1 above)
Comment (NucciTheBoss) [reply lib/charms/hpc_libs/v0/slurm_ops.py:672]: `self.var_lib_path.chmod(0o755)` is used here because of the perturbations between each of the Slurm daemon packages. `/var/lib/slurm` is created with `0o755` file permissions for `slurmctld`, `slurmd`, and `slurmdbd`, but not `slurmrestd`.

`parents=True` will ensure that `/var/lib/slurm` is created on `slurmrestd` machines, but the file permissions mode `0o755` will only be applied to the `StateSaveLocation` directory `/var/lib/slurm/checkpoint`, not `/var/lib/slurm`. Its mode will be `0o775` which is looser than I would like, so that's why `self.var_lib_path.chmod(0o755)` is there. It's more a stop-gap to ensure `/var/lib/slurm` looks and feels the same on all the Slurm nodes. 

If we end up using `var_lib_path` in more places than just in `_AptManager` for creating the `StateSaveLocation`, and `_JWTKeyManager` for storing the JWT key in `StateSaveLocation`, then I'd agree that we should move managing the permissions for `var_lib_path` outside of `_create_state_save_location`.
